File: src/services/recommendations.py
from src.services.market_data import fetch_batch, STOCK_UNIVERSE, ETF_UNIVERSE, format_market_cap
from src.services.risk_profile import get_allocation


def generate_recommendations(risk_score: int, profile_label: str) -> dict:
    allocation = get_allocation(profile_label)

    all_stocks = fetch_batch(STOCK_UNIVERSE, cached_only=True)
    all_etfs = fetch_batch(ETF_UNIVERSE, cached_only=True)

    scored_stocks = [_score_instrument(s, risk_score, profile_label) for s in all_stocks if s]
    scored_etfs = [_score_instrument(e, risk_score, profile_label) for e in all_etfs if e]

    scored_stocks.sort(key=lambda x: x["match_score"], reverse=True)
    scored_etfs.sort(key=lambda x: x["match_score"], reverse=True)

    recs = scored_stocks[:8] + scored_etfs[:7]
    recs.sort(key=lambda x: x["match_score"], reverse=True)

    return {
        "profile_label": profile_label,
        "risk_score": risk_score,
        "allocation": {
            "stocks": allocation["stocks"],
            "bonds": allocation["bonds"],
            "cash": allocation["cash"],
        },
        "recommendations": recs,
    }


def _score_instrument(data: dict, risk_score: int, profile_label: str) -> dict:
    score = 50
    beta = data.get("beta")
    pe = data.get("pe_ratio")
    div_yield = data.get("dividend_yield")
    year_chg = data.get("year_change")

    if beta is not None:
        if profile_label in ("Very Conservative", "Conservative"):
            if beta < 0.8:
                score += 20
            elif beta < 1.0:
                score += 10
            elif beta > 1.3:
                score -= 15
        elif profile_label in ("Growth", "Aggressive"):
            if beta > 1.2:
                score += 15
            elif beta > 1.0:
                score += 10
            elif beta < 0.7:
                score -= 10

    if div_yield is not None:
        if profile_label in ("Very Conservative", "Conservative"):
            if div_yield > 3.0:
                score += 15
            elif div_yield > 1.5:
                score += 8
        elif profile_label in ("Growth", "Aggressive"):
            if div_yield > 4.0:
                score -= 5

    if pe is not None and pe > 0:
        if profile_label in ("Very Conservative", "Conservative"):
            if pe < 15:
                score += 10
            elif pe > 35:
                score -= 10
        elif profile_label in ("Growth", "Aggressive"):
            if pe < 30:
                score += 5

    if year_chg is not None:
        if year_chg > 20 and risk_score >= 7:
            score += 10
        elif year_chg < -10 and risk_score <= 3:
            score -= 10

    rec = data.get("recommendation", "")
    if rec in ("buy", "strong_buy"):
        score += 10
    elif rec == "sell":
        score -= 15

    score = max(10, min(99, score))

    risk_level = "Low"
    if beta and beta > 1.3:
        risk_level = "High"
    elif beta and beta > 1.0:
        risk_level = "Medium"

    reason = _build_reason(data, profile_label, risk_level)

    return {
        "symbol": data["symbol"],
        "name": data["name"],
        "asset_type": data.get("asset_type", "Stock"),
        "sector": data.get("sector", "N/A"),
        "price": round(data.get("price", 0), 2),
        "match_score": score,
        "risk_level": risk_level,
        "reason": reason,
        "pe_ratio": round(pe, 2) if pe else None,
        "dividend_yield": div_yield,
        "beta": round(beta, 2) if beta else None,
        "year_change": year_chg,
    }


def _build_reason(data: dict, profile_label: str, risk_level: str) -> str:
    parts = []
    beta = data.get("beta")
    div_yield = data.get("dividend_yield")

    if profile_label in ("Very Conservative", "Conservative"):
        if beta and beta < 0.8:
            parts.append("Low volatility")
        if div_yield and div_yield > 2.5:
            parts.append(f"Strong {div_yield:.1f}% dividend yield")
        if not parts:
            parts.append("Established company with stable track record")
    elif profile_label == "Moderate":
        parts.append("Balanced risk-return profile")
        if div_yield and div_yield > 1.5:
            parts.append(f"{div_yield:.1f}% dividend")
    else:
        if beta and beta > 1.2:
            parts.append("High growth potential")
        yr = data.get("year_change")
        if yr and yr > 15:
            parts.append(f"Strong momentum (+{yr:.0f}% YoY)")
        if not parts:
            parts.append("Growth-oriented with upside potential")

    return ". ".join(parts)

File: static/js/recommendations.js
let allocChart = null;
let _recsData = null;

async function loadRecommendations() {
    const container = document.getElementById("recs-container");
    const noProfile = document.getElementById("recs-no-profile");

    container.innerHTML = '<div class="loading-spinner"><div class="spinner"></div><p>Generating personalized recommendations...</p></div>';
    noProfile.style.display = "none";

    try {
        const data = await api.get("/api/recommendations");
        renderRecommendations(data);
    } catch (err) {
        container.innerHTML = "";
        noProfile.style.display = "block";
    }
}

function renderRecommendations(data) {
    const container = document.getElementById("recs-container");
    _recsData = data;

    const allocHtml = `
        <div class="recs-top">
            <div class="alloc-card chart-card">
                <h3>Recommended Allocation</h3>
                <div class="alloc-profile">
                    <span class="badge badge-${data.profile_label.toLowerCase().replace(" ", "-")}">${data.profile_label}</span>
                    <span class="text-muted">Risk Score: ${data.risk_score}/10</span>
                </div>
                <canvas id="chart-alloc" width="250" height="250"></canvas>
            </div>
            <div class="alloc-summary chart-card">
                <h3>Portfolio Split</h3>
                <div class="alloc-row"><div class="alloc-dot" style="background:#6366f1"></div>Stocks ${helpIcon("allocation_stocks")}<span>${data.allocation.stocks}%</span></div>
                <div class="alloc-row"><div class="alloc-dot" style="background:#22c55e"></div>Bonds ${helpIcon("allocation_bonds")}<span>${data.allocation.bonds}%</span></div>
                <div class="alloc-row"><div class="alloc-dot" style="background:#eab308"></div>Cash ${helpIcon("allocation_cash")}<span>${data.allocation.cash}%</span></div>
                <p class="alloc-note">Based on your ${data.profile_label.toLowerCase()} risk profile, we recommend this allocation to balance growth and stability.</p>
            </div>
        </div>`;

    const filterTabs = `
        <div class="rec-tabs" style="display:flex;align-items:center;justify-content:space-between;flex-wrap:wrap;gap:8px;">
            <div style="display:flex;gap:4px;">
                <button class="rec-tab active" onclick="filterRecs('all', this)">All (${data.recommendations.length})</button>
                <button class="rec-tab" onclick="filterRecs('Stock', this)">Stocks</button>
                <button class="rec-tab" onclick="filterRecs('ETF', this)">ETFs</button>
            </div>
            <button class="bundle-buy-btn" onclick="buyRecommendationsBundle()">
                <svg viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2"><path d="M12 2v20M17 5H9.5a3.5 3.5 0 0 0 0 7h5a3.5 3.5 0 0 1 0 7H6"/></svg>
                Buy All as Bundle
            </button>
        </div>`;

    const cards = data.recommendations.map((r, idx) => {
        const riskColors = { Low: "var(--green)", Medium: "#eab308", High: "var(--red)" };
        const matchColor = r.match_score >= 75 ? "var(--green)" : r.match_score >= 50 ? "#eab308" : "var(--red)";
        const safeName = (r.name || "").replace(/'/g, "\\'");

        // Build a richer "Why Selected" explanation from available data
        let whyParts = [];
        if (r.match_score >= 80) whyParts.push(`Excellent ${r.match_score}% match for your ${data.profile_label.toLowerCase()} risk profile`);
        else if (r.match_score >= 60) whyParts.push(`Good ${r.match_score}% match for your ${data.profile_label.toLowerCase()} risk profile`);
        else whyParts.push(`${r.match_score}% match for your ${data.profile_label.toLowerCase()} risk profile`);

        if (r.risk_level === "Low" && r.beta != null && r.beta < 1) whyParts.push(`Low volatility (beta ${r.beta.toFixed(2)}) provides stability`);
        if (r.dividend_yield != null && r.dividend_yield > 2) whyParts.push(`${r.dividend_yield.toFixed(2)}% dividend provides passive income`);
        if (r.pe_ratio != null && r.pe_ratio > 0 && r.pe_ratio < 20) whyParts.push(`Reasonable valuation with P/E of ${r.pe_ratio.toFixed(1)}`);
        if (r.year_change != null && r.year_change > 10) whyParts.push(`Strong momentum with ${r.year_change.toFixed(1)}% year-over-year growth`);
        if (r.year_change != null && r.year_change < -10) whyParts.push(`Currently discounted (${r.year_change.toFixed(1)}% YoY) — potential recovery opportunity`);

        const whyDetailHtml = whyParts.length > 0
            ? `<div class="rec-why-detail" id="rec-detail-${idx}" style="display:none;">
                <div class="rec-why-title">Why ${r.symbol} Was Selected</div>
                <div class="rec-why-text">${whyParts.join(". ")}.</div>
                <div class="rec-why-metrics">
                    ${r.pe_ratio != null ? `<div class="rec-why-metric"><span>P/E</span><strong>${r.pe_ratio.toFixed(1)}</strong></div>` : ""}
                    ${r.beta != null ? `<div class="rec-why-metric"><span>Beta</span><strong>${r.beta.toFixed(2)}</strong></div>` : ""}
                    ${r.dividend_yield != null ? `<div class="rec-why-metric"><span>Div Yield</span><strong>${r.dividend_yield.toFixed(2)}%</strong></div>` : ""}
                    ${r.year_change != null ? `<div class="rec-why-metric"><span>1Y Change</span><strong style="color:${r.year_change >= 0 ? 'var(--green)' : 'var(--red)'}">${r.year_change >= 0 ? '+' : ''}${r.year_change.toFixed(1)}%</strong></div>` : ""}
                </div>
               </div>`
            : "";

        return `
        <div class="rec-card" data-asset-type="${r.asset_type}" data-symbol="${r.symbol}" data-stock-name="${(r.name||"").replace(/"/g,'&quot;')}" data-stock-price="${r.price}" onclick="toggleRecDetail(${idx})">
            <div class="rec-card-header">
                <div>
                    <div class="rec-symbol">${r.symbol}</div>
                    <div class="rec-name">${r.name}</div>
                </div>
                <div class="rec-match" style="color:${matchColor}">${r.match_score}% match ${helpIcon("match_score")}</div>
            </div>
            <div class="rec-badges">
                <span class="badge" data-help="risk_level" style="background:${riskColors[r.risk_level]}20;color:${riskColors[r.risk_level]}">${r.risk_level} Risk</span>
                <span class="badge" style="background:var(--bg);color:var(--text-muted)">${r.asset_type}</span>
                <span class="badge" style="background:var(--bg);color:var(--text-muted)">${r.sector}</span>
            </div>
            <div class="rec-metrics">
                <div><span class="metric-label">Price</span><span class="metric-value">${fmt(r.price)}</span></div>
                <div><span class="metric-label">P/E ${helpIcon("pe_ratio")}</span><span class="metric-value">${r.pe_ratio != null ? r.pe_ratio.toFixed(1) : "—"}</span></div>
                <div><span class="metric-label">Div Yield ${helpIcon("dividend_yield")}</span><span class="metric-value">${r.dividend_yield != null ? r.dividend_yield.toFixed(2) + "%" : "—"}</span></div>
                <div><span class="metric-label">Beta ${helpIcon("beta")}</span><span class="metric-value">${r.beta != null ? r.beta.toFixed(2) : "—"}</span></div>
            </div>
            <div class="rec-reason">${r.reason}</div>
            ${whyDetailHtml}
            <div class="rec-actions" style="display:flex;gap:8px;margin-top:10px;" onclick="event.stopPropagation()">
                <button class="btn btn-sm btn-primary" onclick="openAddHoldingModal('${r.symbol}','${safeName}',${r.price})" title="Add to portfolio">+ Portfolio</button>
                <button class="btn btn-sm" onclick="addToWatchlistFromDetail('${r.symbol}','${safeName}')" title="Add to watchlist">+ Watch</button>
                <button class="btn btn-sm" onclick="navigateToStock('${r.symbol}')" title="View details">📈 Details</button>
            </div>
        </div>`;
    }).join("");

    container.innerHTML = allocHtml + filterTabs + `<div class="recs-grid" id="recs-grid">${cards}</div>`;

    renderAllocChart(data.allocation);
}

function renderAllocChart(alloc) {
    const ctx = document.getElementById("chart-alloc");
    if (!ctx) return;
    if (allocChart) allocChart.destroy();
    allocChart = new Chart(ctx, {
        type: "doughnut",
        data: {
            labels: ["Stocks", "Bonds", "Cash"],
            datasets: [{ data: [alloc.stocks, alloc.bonds, alloc.cash], backgroundColor: ["#6366f1", "#22c55e", "#eab308"], borderWidth: 0 }],
        },
        options: {
            responsive: true, cutout: "60%",
            plugins: { legend: { display: false } },
        },
    });
}

function filterRecs(type, btn) {
    document.querySelectorAll(".rec-tab").forEach((t) => t.classList.remove("active"));
    btn.classList.add("active");
    document.querySelectorAll(".rec-card").forEach((c) => {
        c.style.display = (type === "all" || c.dataset.assetType === type) ? "" : "none";
    });
}

function buyRecommendationsBundle() {
    if (!_recsData || !_recsData.recommendations || _recsData.recommendations.length === 0) {
        if (typeof showToast === "function") showToast("No recommendations to buy", "info");
        return;
    }
    const stocks = _recsData.recommendations.map(r => ({
        symbol: r.symbol,
        name: r.name,
        price: r.price,
        allocation_pct: r.match_score,
    }));
    buyStockBundle(stocks);
}

function toggleRecDetail(idx) {
    const el = document.getElementById(`rec-detail-${idx}`);
    if (!el) return;
    el.style.display = el.style.display === "none" ? "block" : "none";
}
